selftest/selftest.py

```diff
@@ -854,19 +879,160 @@ def parse_azure_nvme_id_version(azure_nvme_id_version_output: str) -> str:
+
+    @classmethod
+    def enumerate_interfaces(cls) -> Dict[str, NetworkInterface]:
+        """Retrieve all Ethernet interfaces on the system."""
+        interfaces: Dict[str, NetworkInterface] = {}
+        interface_names = [
+            interface
+            for interface in os.listdir(SYS_CLASS_NET)
+            if os.path.exists(os.path.join(SYS_CLASS_NET, interface, "device"))
+        ]
+
+        for interface_name in interface_names:
+            sys_path = Path(SYS_CLASS_NET, interface_name)
+            udev_properties = cls.query_udev_properties(interface_name)
+            driver_path = Path(sys_path, "device", "driver")
+            if not driver_path.is_symlink():
+                logger.debug(
+                    "ignoring interface %s without driver symlink", interface_name
+                )
+                continue
+
+            link = os.readlink(driver_path)
+            driver = os.path.basename(link)
+            mac = (sys_path / "address").read_text().strip()
+            ipv4_addrs = cls.get_ipv4_addresses(interface_name)
+            interfaces[interface_name] = NetworkInterface(
+                name=interface_name,
+                driver=driver,
+                mac=mac,
+                ipv4_addrs=ipv4_addrs,
+                udev_properties=udev_properties,
+            )
+
+        return interfaces
+
+    @staticmethod
+    def get_ipv4_addresses(interface: str) -> List[str]:
+        """Get the IPv4 addresses of a given network interface using `ip addr`."""
+        try:
+            result = subprocess.run(
+                ["ip", "-4", "addr", "show", interface],
+                stdout=subprocess.PIPE,
+                stderr=subprocess.PIPE,
+                text=True,
+                check=True,
+            )
+
+            ipv4_addresses = re.findall(r"inet (\d+\.\d+\.\d+\.\d+)", result.stdout)
+            return ipv4_addresses
+        except subprocess.CalledProcessError as error:
+            logger.error("failed to get IPv4 address for %s: %r", interface, error)
+            raise
+
+    @staticmethod
+    def query_udev_properties(interface: str) -> Dict[str, str]:
+        """Query all udev properties for a given interface using udevadm."""
+        try:
+            result = subprocess.run(
+                [
+                    "udevadm",
+                    "info",
+                    "--query=property",
+                    f"--path={SYS_CLASS_NET}/{interface}",
+                ],
+                stdout=subprocess.PIPE,
+                stderr=subprocess.PIPE,
+                text=True,
+                check=True,
+            )
+            properties: Dict[str, str] = {}
+            for line in result.stdout.splitlines():
+                if "=" in line:
+                    key, value = line.split("=", 1)
+                    properties[key] = value
+            return properties
+        except subprocess.CalledProcessError as e:
```

Comment:
I thought pylint would complain about variable name (e) being too short?

Reply:
apparently not, fixed!